### PR TITLE
[WIP] "Checks" (e.g. preconditions/postconditions) as the basis of module testing

### PR DESCRIPTION
**Pull request by apparentlymart**
In Terraform v0.15 we published the first official iteration of experimental support for an integrated automated integration test harness for Terraform modules. (Before that there were some _unofficial_ iterations I did on my own, but that's another story.)

Although we had to take a break from it for a while to focus efforts on the v1.0 release and some other things, I've recently been working with @korinne on figuring out what we might try next.

One particular area we knew was lacking in the first experiment was the use of a weird built-in provider as a substitute for a first-class language syntax for writing test assertions, and indeed that was echoed in a lot of feedback we recieved. While that's certainly not the only opportunity for improvement, it's the focus of this particular pull request!

## "Checks" as the basis of testing

Terraform v1.2 introduced preconditions and postconditions as two kinds of author-defined assertions, which allow a module to "self-check". That is a useful feature in its own right as a way to give better feedback when things don't work out, but it's _also_ a plausible first-class syntax for describing expected outcomes for tests.

Specifically, I imagine two different complementary uses of custom checks in testing:
- Checks written inside the module itself can run both in the automated testing   environment and at runtime in a real configuration, increasing the return   on the investment of writing those checks in the first place.
- However, some checks will probably make sense _only_ in automated testing,   in which case it's reasonable to write the tests in the root modules that define the testing scenarios. This would use the same syntax, but those checks would run only in the test harness, and so wouldn't benefit "real" use of the module.

Because our initial focus for preconditions and postconditions was just to halt with an error message on failure during a normal run, we had only basic support for keeping track of the check outcomes throughout a run.

An automated test harness needs some higher-fidelity information in order to generate useful results:
- When everything passes, it should be able to produce a manifest of all of   the checks so that the author can see the difference between a test executing and succeeding vs. a test not running at all for some reason.
- When encountering an error, we should be able to report on what subset of checks we ran before the error halted execution, which objects failed their checks, and which checks we didn't visit at all. That implies that we need to know a full set of _expected_ checks, even if we end up not reaching all of them due to an early exit.

The bulk of this PR, then, is an expansion of the earlier work on conditions into a new first-class concept I'm calling "checks", which is intentionally a more general term because I'm anticipating other kinds of checks coming under this heading in the future, beyond preconditions and postconditions.

In order to meet the additional requirements above, the main change here is that we begin by analyzing the static configuration to learn about all of the configuration objects that have associated checks (currently just resources and output values), pre-record them as "unknown" in the check state, and then gradually update them with known statuses as we make progress walking the graph.

The outcome then is a full catalog of all of the checkable configuration objects, even if some of them weren't reachable due to an error, and then detailed statuses associated with the instances of each one where possible.

## Configuration Objects vs. Dynamic Objects

In my first iteration of this I tried to just pre-index all of the checks that the existing preconditions/postconditions mechanism might generate, but that ran into the trouble that we were tracking checks exclusively by resource instance address and absolute output value address, which means that we need to resolve all of the `count` and `for_each` expressions in the configuration before we can know what will exist, and an upstream error during planning can mean that certain objects won't get "expanded" (instances decided) at all.

As a compromise then, I've introduced here a distinction between "configuration objects" and "dynamic objects". To understand the difference, here's a contrived example:

```hcl
resource "null_resource" "example" {
  count = 2

  lifecycle {
    postcondition {
      condition     = self.id != ""
      error_message = "The ID is empty, somehow."
    }
  }
}
```

This resource block has at least one condition defined, which is the first requirement for an object to participate in the idea of "checks".

The resource block itself, addresed as `null_resource.example`, is a static _configuration object_ which we can resolve just by looking at the configuration, without evaluating any expressions.

During the graph walk, we expand this into two resource instances `null_resource.example[0]` and `null_resource.example[1]`, both of which are examples of _dynamic checkable objects_.

All of the dynamic objects associated with a particular configuration object are always of the same type and have exactly the same set of conditions.

If we manage to get far enough during planning to "expand" an object, we'll immediately associate the dynamic objects with their configuration object and record all of the dynamic object check statuses as "unknown". Then Terraform Core will visit each instance in turn, recording the actual check outcome for each.

If the plan walk halts before expanding a configuration object, we still know that the configuration object as a whole has status "unknown", and so we can still report that in the test results to meet the requirement of being explicit about what we didn't check.

## Individual checks are a module implementation detail

In the current preconditions/postconditions design we track the status of individual conditions internally, but the small machine-readable data export surface we implemented via `terraform show -json PLANFILE` hides that detail and instead just reports results by the dynamic object they belong to.

I ended up leaning in to that here and making that fundamental in the externally-visible model: internally during a run we do still track the status of each individual check, but we discard that detail on exit from Terraform Core and instead the model is whether a particular _checkable object_ passed or failed.

If an object failed, it will typically have one or more failure messages associated with it, for each check that failed, but without any explicit machine-readable link between the message and the condition that generated it.

My rationale here is that multiple conditions associated with an object is primarily a syntactic convenience for a module author to conveniently write down multiple independent conditions that must all hold (implied logical AND), with a different error message for each. A module author can freely change the number and ordering of the individual checks without that being externally visible, but the object the checks are associated with has a durable identity that we can use to compare the results from one run to the results from another.

## Checks in the State

The current implementation of preconditions/postconditions focuses only on a single run, and so the condition results are discarded as soon as the apply step concludes.

As a building block for other work, including reporting results from a test harness, I've extended that model here to record check results as part of the state snapshot generated at the end of the apply step. That's particularly interesting if the apply failed partway through due to an error, because in that case Terraform has already partially updated infrastructure and so it can serve as some additional feedback on what checks already ran and passed, which checks failed, and which objects didn't get checked at all.

This also gives Terraform a "memory" of what passed and failed in the previous run, and so in principle we could add a new section to the plan output reporting that a previously-failing check has now been fixed, giving some positive feedback:

```
- null_resource.example[0] checks are now fixed!
```

I've avoided making any significant changes to any "main" UI here though, since this is just some exploration for now. Instead, I added an entirely new experimental `terraform checks` command that just reads the check results from the most recent state snapshot and displays them:

```
$ terraform checks -all
❌ data.javascript.foo failed
  - Reticulated splines are still too spikey.
✅ null_resource.a passed
✅ null_resource.b[0] passed
✅ null_resource.b[1] passed
✅ null_resource.b[2] passed
✅ null_resource.b[3] passed
✅ null_resource.b[4] passed
✅ null_resource.c passed
```

This is of course just placeholder UI. I don't know if it's really even useful to have a dedicated command to report checks, but it's a convenient way to play with the mechanisms introduced in this PR so far.

## What about testing, then?

Nothing I've said so far here has talked directly about testing! Introducing this explicit model for checks was already a big enough change that I wanted to let this sit and discuss it a bit for a while before going much further, but at a high level my idea is that this concept of checks will replace the use of instances of the test assertion resource in the placeholder provider in the current experiment, allowing us to fit our test results approximately to the de-facto standard "JUnit-style" report structure like this:

* Test scenarios (the directories under `tests`) remain analogous to test suites.
* Terraform modules in each test configuration are analogous to "classes".
* The checks themselves are analogous to "test cases".
* Check statuses map to JUnit test case outcomes as follows:
    * Pass = Pass
    * Fail = Fail
    * Error = Error (in JUnit, used to report an exception)
    * Unknown = Skipped

A particular test scenario would for now therefore be considered successful if it can plan, apply, and destroy without errors _and_ all of its declared checks pass after the apply step.

I'm also planning to look into offering a way to allow authors to describe more customized test scenarios, which can describe a sequence of operations rather than just a single plan/apply/destroy.

As part of that, I'm hoping to support declaring _expected failures_, where the test author can specify the addresses of a particular set of objects that should have the fail status in order for the test to be considered successful, and therefore authors will be able to test that the conditions themselves are working in addition to testing the main functionality of the module. Since
all of the dynamic objects have a user-visible address, we'll have a ready-to-use syntax for specifying the objects whose checks are supposed to fail in a particular scenario.

## More to come!

This is just an incremental step on the way. I don't know if anything here is actually good to merge and move forward with yet, but hopefully at least some parts of it are. For now though, the main idea is to just stop and come up for air and discuss what I've worked on so far.

### CODE REVIEW

**Comment by apparentlymart** [comment]
Thanks for the feedback, @alisdair!

I must admit I was a little unsure as to what compatibility commitments applied to the JSON serialization of conditions and I was meaning to ask you about that specifically but I somehow lost the note to do so along the way so I'm glad you picked up on it anyway! If we do have some room to phase out `condition_results` in favor of `check_results` then that does indeed seem like a nice way to navigate this, so that we can have them both coexisting for a little while but not be stuck emitting (and storing, for Terraform Cloud) the redundant copy of the same data for all future releases.

The emojis in `terraform checks` are indeed unprecedented and I didn't actually want to use emojis there but when I went looking for the non-emoji Unicode checkmark and similar I found that a number of characters that were originally just "dingbats" have retroactively become emojis in later Unicode versions and so I just went ahead and used emojis throughout because my primary concern was to see the data at all, rather than worry too much about the UI of a command that might not actually end up being a real thing in the end anyway. (It's still not clear to me that it's useful as a separate command if we integrate checks more carefully into the main workflow.) With that said though, I do share your reluctance about beginning to use emoji here because of the various accessibility and terminal compatibility unknowns it would bring into play, and so if we _do_ end up having a command like this I expect we'd find a different compromise that is hopefully still relatively concise but also more broadly accessible.

I think my next step for this will be to be to wind back to the beginning of this branch and try for a more direct path of implementation, without the design meandering and subsequent inconsistencies that are here right now. I intend it to be architecturally basically the same but in particular I've noticed a few ways I wasn't consistent about naming or organization that I want to clean up, along with writing some more complete test coverage on my second attempt. But before I do that I'm also going to take a first pass at wiring this up to `terraform test` and see how that feels, because I want to make sure there aren't any hidden surprises there I haven't thought of yet. For the moment then, I'll leave this as a draft until I've done the revisions that will make this ready for a final review.

Thanks!

**Comment by apparentlymart** [comment]
This PR originally had a bunch of changes in it that were not directly related to checks but that I relied on to make the checks design work. In the meantime I split them all out into separate PRs which are now landed, so I've rebased this one to remove all of the distracting unrelated stuff.

I am still planning to rework this a bunch to tidy up some messes that resulted from me exploring this gradually and learning along the way rather than having a well-formed design to start with. My intent is that the new version will be functionally-equivalent from an external viewpoint, but the internal concepts will probably move around a little and have some slightly different names on my next pass, just because I know a few things are already inconsistent/confusing in the current state.

I am also planning to totally remove the `terraform checks` command for my next pass, since I've reached the conclusion that it isn't really super useful in its current form and it'd be better to invest in making checks more visible in the normal plan/apply workflow. I'm not planning to introduce any UI changes to the plan/apply workflow in this PR, and will instead focus on this generating the necessary data in the plan and state so that a _later_ PR can make use of it in the UI layer.

File: internal/checks/status.go
Comment (alisdair) [inline internal/checks/status.go:71]: Do we consider a marked condition result as panic-worthy? I would have expected us to unmark the value before comparison here, although I don't have a solid rationale for why.

File: internal/command/jsonchecks/checks.go
Comment (alisdair) [inline internal/command/jsonchecks/checks.go:73]: I really like this idea (and the corresponding note on the old JSON output for conditions)!

File: internal/plans/planfile/tfplan.go
Comment (alisdair) [inline internal/plans/planfile/tfplan.go:174]: Nilling these empty maps feels a bit tricky to me, only because elsewhere in this changeset we logically distinguish between an `addrs.Map`'s `Elems` being `nil` vs being an empty map (in `AggregateCheckStatus`). There's no bug here as far as I can tell, but I wonder if it would be clearer to be consistent by leaving this as an empty map.

(Noting also that the documentation for `addrs.Map` says that "callers MUST NOT modify this map directly", which also makes me think that this nilling is a little bit suspicious.)

File: internal/states/checks.go
Comment (alisdair) [inline internal/states/checks.go:77]:
```suggestion
// NewChecks constructs a new states.CheckResults object that is a snapshot of the
```

File: internal/states/checks.go
Comment (alisdair) [inline internal/states/checks.go:74]: A very wooly thought about this structure: in future we might want to add more failure information for an individual check, such as a label, or a condition expression representation. I think that would mean changing this into a set of objects, like `Failures []CheckResultFailure`.

I think that in our JSON serialization we already make space for this by creating an object per failure with a `"message"` key. Is it worth doing the same thing here for ease of migration alter, or is it too early?

**Review by alisdair — APPROVED** [review]
This is great! I smoke tested it locally and found everything to work as expected, including the JSON output.

I have some minor thoughts inline, but nothing blocking merge.

File: internal/checks/status.go
Comment (apparentlymart) [reply internal/checks/status.go:71]: I think my intent here had been to make it the caller's responsibility to reduce the value to the very specific domain of this function before calling it, because the caller has more context about what's going on and so is in a better position to handle e.g. reporting an actionable error if something is marked but shouldn't be, or deciding that it's okay to just quietly drop the marks.

Currently that logic lives in `eval_conditions.go` and it does just silently drop the marks right before calling this function, but to me it feels better to leave that decision somewhere more prominent than hiding inside this function so that a reader of `evalCheckRule` can clearly see both _that_ we are considering sensitive values and _how_ we are considering them.

It is a position weakly held, however.

File: internal/plans/planfile/tfplan.go
Comment (apparentlymart) [reply internal/plans/planfile/tfplan.go:174]: I believe what motivated this is that the file format for plans cannot actually distinguish between an empty map and a nil map -- we serialize as protobuf which has no explicit serialization for "map of zero elements" separate from there being no map set at all -- and so this extra special case is to pick one specific representation of "no results".

I must admit I don't recall if there was a reason why it was better for the canonical representation to be `nil` rather than being a zero-length non-nil map; I could try it and see if it's practical to change the _other_ end of this so that the "elems" are _never_ `nil` in this case. I think the main thing is that we need to decide one way or the other because we cannot possibly preserve the difference through protobuf serialization.

File: internal/states/checks.go
Comment (apparentlymart) [reply internal/states/checks.go:74]: I did indeed intentionally make the JSON serialization have room for expansion, because that's the only part that will become a compatibility constraint moving forward.

Honestly if this had come about in a different order where I knew I'd be prioritizing the JSON output rather than the `terraform test` command I likely would've built this as a slice of some struct, but while I was revising to change the focus I had concluded that it wouldn't be a big deal to change this internal model later if we need to.

With that said, this also probably isn't a big deal to change proactively now. I think what I will do is land this PR with it implemented this way and then I'll make a smaller follow-up PR that deals with this refactoring which we can safely land at any point in the v1.3 maintenance period without changing any externally-visible behavior. (and similarly for your other feedback bits above, which I think are also worth exploring.)

Comment (apparentlymart) [reply internal/states/checks.go:74]: I did take a quick attempt at this just after merging and found that it's a little trickier than I hoped because it really ought to be extended in both the `checks` package and the `states` package in a similar way, and I want to spend a little more time pondering whether it's useful to share a struct type between both of those or whether it's worth keeping them separate, and other similar internal API design questions.

I do intend to pick this up again later but Friday-afternoon brain is insufficient for this sort of design work so I'll just keep it in mind for exploring in the near future. Thanks for pointing it out!

**Pull request merged**
